feat: oracle node

matt-user opened the pull request:
Huge shoutout to @segfault-magnet for all the help
## Type of change

<!--Delete points that do not apply-->

- New feature

## Changes

The following changes have been made:

- Create an oracle node to interact with the smart contract

## Related Issues

<!--Delete everything after the "#" symbol and replace it with a number. No spaces between hash and number-->

Closes #165

Braqzen commented on oracle/packages/node/.env.example:
```diff
@@ -0,0 +1,4 @@
+API_URL="https://min-api.cryptocompare.com/data/price?fsym=ETH&tsyms=USD&api_key=<your api key here>"
+ORACLE_CONTRACT_ID=0x1bc1332bff931c3cfc9752fe10fe30e1ed3c41f77101f7d144c8183355e5590e
+WALLET_SECRET=0x0000000000000000000000000000000000000000000000000000000000000001
+FUEL_PROVIDER_URL=http://localhost:4000/graphql
```
```suggestion
FUEL_PROVIDER_URL=http://localhost:4000/graphql

```

Braqzen commented on oracle/README.md:
````diff
@@ -34,13 +37,36 @@ oracle/
 TODO: need UI for this to be relevant
 
+### Node
+In order to run the Oracle node make sure you are in the root of this project i.e `/path/to/oracle/<you are here>`
+
+Copy the `.env.example` file into a new file called `.env`.  Then insert your api key into the `API_URL`
+
+**_Note:_** You do not need an api key to run any tests
+
+
+Then start a local fuel-core instance
+```bash
+fuel-core run --chain packages/node/.chainConfig.json
+```
+
+Then deploy an instance of the Oracle contract
+```bash
+forc deploy --path packages/contract --url localhost:4000 --unsigned
+```
+
+Finally start the Oracle node
+```bash
+cargo run
+```
````
```suggestion

The node can be started by executing the following steps:

1. Change into the `Oracle` directory

```bash
cd <path>/sway-applications/oracle/<you are here>
```

2. Copy and paste the `.env.example` file into a new file called `.env`

```bash
cp node/.env.example node/.env
```

3. In the newly copied `.env` file there is a variable `API_URL` which ends with `<your api key here>`. This section should be replaced with your API key.

4. Start `fuel-core`

```bash
fuel-core run --chain packages/node/.chainConfig.json
```

5. Deploy the Oracle contract

```bash
forc deploy --path packages/contract --url localhost:4000 --unsigned
```

6. Start the Oracle node

```bash
cargo run
```
```

This looks weird in the suggestion on github. I suggest checking this out locally to make it look good.

Points to note: 

1. This does not tell the reader anything about acquiring an API key. That is confusing. "Put this mysterious object that I've told you nothing about here". How would a user know what to do to acquire such an object?
2. Steps 4 and 5 seem odd to me. If I only want to "run a node" then why am I required to run some `fuel-core` and also deploy a contract? Moreover, `deploy` is a plugin which they must install separately so that might error for them.

The readme requires a bit more info

matt-user replied:
Good suggestions I'll add more info/rework the readme

Braqzen commented on oracle/README.md:
````diff
@@ -34,13 +37,36 @@ oracle/
 
 ```bash
-forc test
+cargo test
````
Since a workspace is being used I hope this ensures that all the tests can be run. I'm getting a bit confused with the CI using `forc test`, the later releases deprecating that command and using `cargo test` (while we still have not bumped the repo to reflect this) and now this uses `cargo test` too.

matt-user replied:
Cargo test ensures all tests are run.  However, on a second look I don't even think the CI is running the oracle node tests because Cargo test is never called.

Braqzen commented on oracle/packages/node/.chainConfig.json:
```diff
@@ -0,0 +1,30 @@
+{
```
Seems like magic to me, should this be documented somewhere (if it isn't)?

matt-user replied:
I'll add it to the readme in the explanation of the `fuel-core ...` command.  Its purpose is to configure the local fuel node with a wallet `0x09c...` and gives the wallet a lot of the base asset.  I'll add documentation about the  specified wallet the secret_key in `.env.example`

Braqzen commented on oracle/packages/node/src/lib.rs:
```diff
@@ -0,0 +1,103 @@
+/// - `period` - duration to wait before fetching and updating the price for the oracle
+/// - `price_fetcher` - fetches the latest price for an asset
+pub fn spawn_oracle_updater_job(
+    price_updater: impl PriceUpdater + Send + 'static,
```
Oh I can tell that this was fun to write haha

Braqzen commented on oracle/packages/node/.env.example:
```diff
@@ -0,0 +1,4 @@
+API_URL="https://min-api.cryptocompare.com/data/price?fsym=ETH&tsyms=USD&api_key=<your api key here>"
+ORACLE_CONTRACT_ID=0x1bc1332bff931c3cfc9752fe10fe30e1ed3c41f77101f7d144c8183355e5590e
+WALLET_SECRET=0x0000000000000000000000000000000000000000000000000000000000000001
```
Does this secret key matter; can it be set to anything; must it be a real secret key from `forc-wallet`?

Braqzen commented on oracle/packages/node/tests/functions/mod.rs:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    pub fn invocations(&self) -> Arc<Mutex<Vec<Invocation>>> {
+        Arc::clone(&self.invocations)
```
oof, so I'm aware of this area however I've not delved into it so it's a good idea to have an experienced Rust dev take a look at the implementation to see if this makes sense (not just this line but in general)

matt-user replied:
Yeah I'll ping the rustceans in the sdk team.  @segfault-magnet wrote most of the code  :)

segfault-magnet replied:
It's used only in tests. It's a poor man's [spy](https://martinfowler.com/bliki/TestDouble.html):

> Spies are stubs that also record some information based on how they were called. One form of this might be an email service that records how many messages it was sent.

Since the spawn oracle updater fn requires ownership over the price fetcher and price updater, you need some way of giving away the price updater but still having a way of spying on how it is being used for testing purposes. And it mustn't make the updater `!Send`. 

A couple of ways to go about it, one of those is for the spy to hold an inner collection of invocations that is behind a reference count -- i.e. has shared ownership. The Mutex part derives from the 'mustn't make the updater `!Send`' part.

Or you can use a library for generating testing doubles. Since I haven't worked with any such library in Rust, felt quicker to just DIY.

Braqzen commented on oracle/packages/node/tests/functions/run.rs:
```diff
@@ -0,0 +1,85 @@
+#[cfg(test)]
```
How come this needs to be annotated while the reset of the apps in the repo do not?

segfault-magnet replied:
A remnant of when it used to be a unit test. Should be removed since I don't think `test` is active while compiling integration tests.

Hence these tests might not even run with this cfg guard.

Braqzen commented on oracle/Cargo.toml:
```diff
@@ -1,15 +1,12 @@
+    "packages/contract",
+    "packages/node",
+    "packages/utils",
+]
```
```suggestion
]

```

Braqzen reviewed (commented):
Some questions and documentation. After the resolution I'd like to run this myself to see it work.

matt-user commented on APPLICATION_PROGRESS.md:
```diff
@@ -400,6 +400,9 @@ Will move from Apps repo to Libs repo soon
 
 <h3>Oracle Node</h3>
 
+- Feature complete, but could most likely be improved once event monitoring is added to
```
Not in this implementation no.  I originally thought it would from my initial [research](https://docs.chain.link/docs/architecture-request-model/#off-chain-oracle-node)) into oracles, but this is an extremely simple oracle.  Maybe this line should be removed from the readme or say "This oracle could be improved to resemble something closer to chainlink once event monitoring is added"

matt-user commented on oracle/README.md:
```diff
@@ -17,13 +19,14 @@ The project consists of an oracle smart contract and an oracle node which intera
-├── contract/
-|    └── src/main.sw
-|    └── tests/harness.rs
+├── packages/
```
One thing to note is project structure doesn't have to be this way I just took it from [fuel-rs](https://github.com/FuelLabs/fuels-rs/blob/master/Cargo.toml).  We could also do what [fuel-core](https://github.com/FuelLabs/fuel-core/blob/master/Cargo.toml) does. I don't have a strong preference.

segfault-magnet replied:
What we have in the `packages` folder in fuels-rs are really cargo packages. Each folder has a `Cargo.toml` and in it, you can find the `[package]` header detailing info about that cargo package.

[From the rust book](https://doc.rust-lang.org/book/ch07-01-packages-and-crates.html):

> A package is a bundle of one or more crates that provides a set of functionality. A package contains a Cargo.toml file that describes how to build those crates. Cargo is actually a package that contains the binary crate for the command-line tool you’ve been using to build your code. The Cargo package also contains a library crate that the binary crate depends on. Other projects can depend on the Cargo library crate to use the same logic the Cargo command-line tool uses.

Didn't look into whether packages is appropriate here, just wanted to give some insights why we're calling it `packages` in fuels-rs.

matt-user commented:
@Braqzen I added more info to the doc.  I removed the "Node" header as the steps pertain more to the project overall rather than just the node.  I'm not a huge fan of where the env variable info is located.  Maybe it would be better to move that info into the `packages/node` folder and link to it in the readme?

Braqzen commented on oracle/README.md:
```diff
@@ -37,31 +37,57 @@ oracle/
 
-### Node
-In order to run the Oracle node make sure you are in the root of this project i.e `/path/to/oracle/<you are here>`
+The project can be started by executing the following steps:
```
This needs a heading since it looks like it all falls under the UI section defined above

Braqzen commented on oracle/README.md:
````diff
@@ -37,31 +37,57 @@ oracle/
-```bash
-forc deploy --path packages/contract --url localhost:4000 --unsigned
-```
+### Environment Variables
````
This table breaks the flow of the entire document. It should be listed at the very bottom of this section and it should be referenced when going through the steps so that the user can see the steps in sequence followed by being able to scroll / click on the link to the relevant section rather than having a massive chunk of text in between the steps. Also, be consistent with the periods at the end of the sentences.

Braqzen commented on oracle/README.md:
```diff
@@ -37,31 +37,57 @@ oracle/
+| WALLET_SECRET      | Private key of the first deterministic wallet provided by the fuels-rs sdk.  It is also the private key corresponding to the `owner` address specified in the oracle contract's `Forc.toml`.  It is configured to have the maximum amount of the `BASE_ASSET` by the local `fuel-core` instance spun up in the step 4. |
+| FUEL_PROVIDER_URL  | Fuel-core network url normally set as http://localhost:4000/graphql for development |
+
+3. The oracle node relies on an external 3rd-party service to get price information to provide to the oracle contract.  We do not endorse this service neither are we affiliated with them in any way.  We only use the service for demonstration purposes.  If you wish to run the node you can sign-up for a free api key [here](https://www.cryptocompare.com/).  In the newly copied `.env` file there is a ...
```
This is 1 large block of text which mixes commentary and instructions. This should be separated so that the instructions are instructions and any relevant commentary is referenced elsewhere / cut down to be concise.

Braqzen commented on oracle/README.md:
```diff
@@ -37,31 +37,57 @@ oracle/
+
+3. The oracle node relies on an external 3rd-party service to get price information to provide to the oracle contract.  We do not endorse this service neither are we affiliated with them in any way.  We only use the service for demonstration purposes.  If you wish to run the node you can sign-up for a free api key [here](https://www.cryptocompare.com/).  In the newly copied `.env` file there is a ...
+
+    **_Note:_** You do not need an api key to run any tests
```
Again, no note is needed. If it's not mentioned in the tests then the assumption is that it's not needed.

Braqzen commented on oracle/README.md:
````diff
@@ -37,31 +37,57 @@ oracle/
+    fuel-core run --chain packages/node/.chainConfig.json
+    ```
+
+    This will allow us to deploy the oracle contract.  It configures the local `fuel-core` instance with the variables specified in `.chainConfig.json`.  It also initializes the wallet specified in `.chainConfig.json` with the maximum amount of the `BASE_ASSET`.
````
3 separate sentences in 1 point. Try to rewrite it to be as concise as the sentence in the bullet point below. That sentence is great.

Braqzen replied:
Similar comment to the wallet secret in the table. If the rust sdk is mentioned then it should also be hyperlinked so that the UX is good.

digorithm commented on oracle/packages/node/src/lib.rs:
```diff
@@ -0,0 +1,103 @@
+use tokio::task::JoinHandle;
+use tokio::time::sleep;
+
+// Deicmal precision of the asset we are pushing prices to
```
```suggestion
// Decimal precision of the asset we are pushing prices to
```

Braqzen reviewed (dismissed):
eh, I said that I wanted to run this to make sure it works but I cba to get an API key :)

Lgtm so approving.

matt-user commented:
> eh, I said that I wanted to run this to make sure it works but I cba to get an API key :)
> 
> Lgtm so approving.

sounds good, I still have to add the node tests to the ci somehow.  After that I'll merge.

Braqzen commented:
> > eh, I said that I wanted to run this to make sure it works but I cba to get an API key :)
> > Lgtm so approving.
> 
> sounds good, I still have to add the node tests to the ci somehow. After that I'll merge.

Our `build.sh` script might be useful

Braqzen reviewed (commented):
I'm unsure about the latest change for running the tests and splitting up the formatting. That's more of a `misc` issue and should really be separate for organizational purposes, moreover, if we're going in this direction then CI might as well be changed to have a single file which formats and builds rather than duplicating content across 2 files - which adds more credence to making this its own PR and then merging it into here and then merging this into master

matt-user commented:
> I'm unsure about the latest change for running the tests and splitting up the formatting. That's more of a `misc` issue and should really be separate for organizational purposes, moreover, if we're going in this direction then CI might as well be changed to have a single file which formats and builds rather than duplicating content across 2 files - which adds more credence to making this its own PR and then merging it into here and then merging this into master

Yeah I figured we might have to implement the ci changes separately.  I just wanted to get the ci to work correctly w my new changes.  I'm afk rn but I'll make a misc issue and corresponding pr when I get the chance
